url4aac7e6f55_simon0191_tetris3d_tgz-pJ8-Tetris3dJ8/src/vsdk/toolkit/environment/geometry/ParametricBiCubicPatch.java

```diff
@@ -0,0 +1,554 @@
+            M_MATRIX = ParametricCurve.BEZIER_MATRIX;
+        }
+        else if ( this.type == ParametricCurve.HERMITE ) {
+            buildGeometryMatricesXYZ_Hermite();
+            M_MATRIX = ParametricCurve.HERMITE_MATRIX;
+        }
+        else if ( this.type == ParametricBiCubicPatch.FERGUSON ) {
+            buildGeometryMatricesXYZ_Ferguson();
+            M_MATRIX = ParametricCurve.HERMITE_MATRIX;
+        }
+        Mt_MATRIX = new Matrix4x4(M_MATRIX);
+        Mt_MATRIX.transpose();
+        M_Gx_Mt_MATRIX = M_MATRIX.multiply(Gx_MATRIX).multiply(Mt_MATRIX);
+        M_Gy_Mt_MATRIX = M_MATRIX.multiply(Gy_MATRIX).multiply(Mt_MATRIX);
+        M_Gz_Mt_MATRIX = M_MATRIX.multiply(Gz_MATRIX).multiply(Mt_MATRIX);
+        S_MATRIX = new Matrix4x4();
+        Tt_MATRIX = new Matrix4x4();
+        S_MATRIX_DS = new Matrix4x4();
+        Tt_MATRIX_DT = new Matrix4x4();
+    }
+
+    public int getApproximationSteps() {
+        return approximationSteps;
+    }
+
+    public void setApproximationSteps(int n) {
+        approximationSteps = n;
+    }
+
+    public int getType() {
+        return type;
+    }
+
+    public void setType(int type) {
+        type = type;
+    }
+
+    private void buildGeometryMatricesXYZ_Bezier() {
+        double[][] mx = new double[4][4];
+        double[][] my = new double[4][4];
+        double[][] mz = new double[4][4];
+        int i;
+        int j;
+
+        for ( i = 0; i < 4; i++ ) {
+            for ( j = 0; j < 4; j++ ) {
+                Vector3D vp = controlMeshPoints[i][j];
+                mx[i][j] = vp.x;
+                my[i][j] = vp.y;
+                mz[i][j] = vp.z;
+            }
+        }
+        Gx_MATRIX.M = mx;
+        Gy_MATRIX.M = my;
+        Gz_MATRIX.M = mz;
+        //printGeometryMatrices();
+    }
+
+    private void buildGeometryMatricesXYZ_Hermite() {
+        double[][] mx = new double[4][4];
+        double[][] my = new double[4][4];
+        double[][] mz = new double[4][4];
+
+        /* The upper-left 2x2 matrix portion of Gh contains the x coordinates
+           of the four corners of the patch.
+           The upper-right and lower-left 2x2 matrix portion of Gh contains
+           the x coordiates of the tangent vectors four corners of the patch.
+        */
+        int p = 0;
+        int i = 0;
+        for (int j = 0; j < 2; j++) {
+            Vector3D[] vp = contourCurve.getPoint(p);
+
+            mx[i][j] = vp[0].x;
+            my[i][j] = vp[0].y;
+            mz[i][j] = vp[0].z;
+
+            mx[i][j + 2] = vp[2 - j].x;
+            my[i][j + 2] = vp[2 - j].y;
+            mz[i][j + 2] = vp[2 - j].z;
+
+            mx[i + 2][j] = vp[1 + j].x;
+            my[i + 2][j] = vp[1 + j].y;
+            mz[i + 2][j] = vp[1 + j].z;
+
+            Vector3D vn = new Vector3D(0, 0, 0);
+
+            mx[i + 2][j + 2] = vn.x;
+            my[i + 2][j + 2] = vn.y;
+            mz[i + 2][j + 2] = vn.z;
+            p++;
+        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 174:195, 198:220.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.